patches/bess/0012-Dpdk-concurrent-hash-support-pipeline-improvement.patch

```diff
@@ -1,15 +1,59 @@
-From 460ac23d8143f8475a7819aa8ccbbf930f2940c8 Mon Sep 17 00:00:00 2001
-From: Amar Srivastava <amarsri28@gmail.com>
-Date: Wed, 23 Dec 2020 02:33:49 -0800
-Subject: [PATCH] dpdk-hash linking and pipeline improvements
+From 3a29f6b5fd81c4f5517a955e37aa77ad76f3912f Mon Sep 17 00:00:00 2001
+From: Amar Sri <amarsri28@gmail.com>
+Date: Tue, 9 Mar 2021 02:16:36 -0800
+Subject: [PATCH] Exact Match Bulk support
 
 ---
+ core/modules/exact_match.cc    |  25 ++--
  core/modules/wildcard_match.cc | 202 ++++++++++++++++++++++++++-------
  core/modules/wildcard_match.h  |  37 ++++--
  core/utils/cuckoo_map.h        | 146 +++++++++++++++++++++++-
- core/utils/exact_match_table.h |  97 +++++++++++-----
- 4 files changed, 403 insertions(+), 79 deletions(-)
+ core/utils/exact_match_table.h |  92 ++++++++++-----
+ 5 files changed, 407 insertions(+), 95 deletions(-)
 
+diff --git a/core/modules/exact_match.cc b/core/modules/exact_match.cc
+index 1ca0fd4e..5313feea 100644
+--- a/core/modules/exact_match.cc
++++ b/core/modules/exact_match.cc
+@@ -312,7 +312,6 @@ void ExactMatch::setValues(bess::Packet *pkt, ExactMatchKey &action) {
+ void ExactMatch::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
+   gate_idx_t default_gate;
+   ExactMatchKey keys[bess::PacketBatch::kMaxBurst] __ymm_aligned;
+-
+   default_gate = ACCESS_ONCE(default_gate_);
+ 
+   const auto buffer_fn = [&](bess::Packet *pkt, const ExactMatchField &f) {
+@@ -326,18 +325,18 @@ void ExactMatch::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
+ 
+   int cnt = batch->cnt();
+   Value default_value(default_gate);
+-
+-  for (int i = 0; i < cnt; i++) {
+-    bess::Packet *pkt = batch->pkts()[i];
+-    ValueTuple res;
+-    res = table_.Find(keys[i], default_value);
+-    if (res.gate != default_gate) {
+-      /* setting respecive values */
+-      setValues(pkt, res.action);
+-    }
+-    gate_idx_t g = res.gate;
+-    EmitPacket(ctx, pkt, g);
+-  }
++  ValueTuple *res[cnt]; 
++  uint64_t hit_mask = table_.Find(&keys[0],&res[0],cnt);
```

Comment:
The reason I removed the hitmask in my example was it would make an assumption without any assert that cnt is <=64

Reply:
also when simplifying there must be some additional check in `lookup_bulk_data` of cuckoomap to lookup in chunks of 64 at the max in case cnt or n is more than 64

Reply:
bess already check that in pipeline for cnt .

Reply:
Is there a need for hit_mask being exposed all the way up?